src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -414,7 +422,266 @@ public static unsafe void Copy(Array sourceArray, int sourceIndex, Array destina
+            ArgumentNullException.ThrowIfNull(destinationArray);
+
+            if (sourceArray.GetType() != destinationArray.GetType() && sourceArray.Rank != destinationArray.Rank)
+                throw new RankException(SR.Rank_MustMatch);
+
+            ArgumentOutOfRangeException.ThrowIfNegative(length);
+
+            int srcLB = sourceArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(sourceIndex, srcLB);
+            if (sourceIndex - srcLB < 0)
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+            sourceIndex -= srcLB;
+
+            int dstLB = destinationArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(destinationIndex, dstLB);
+            if (destinationIndex - dstLB < 0)
+                throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
+            destinationIndex -= dstLB;
+
+            if ((uint)(sourceIndex + length) > sourceArray.NativeLength)
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+            if ((uint)(destinationIndex + length) > destinationArray.NativeLength)
+                throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
+
+            ArrayAssignType assignType;
+
+            if (sourceArray.GetType() == destinationArray.GetType()
+                || (assignType = CanAssignArrayType(sourceArray, destinationArray)) == ArrayAssignType.SimpleCopy)
+            {
+                MethodTable* pMT = RuntimeHelpers.GetMethodTable(sourceArray);
+
+                nuint elementSize = (nuint)pMT->ComponentSize;
+                nuint byteCount = (uint)length * elementSize;
+                ref byte src = ref Unsafe.AddByteOffset(ref MemoryMarshal.GetArrayDataReference(sourceArray), (uint)sourceIndex * elementSize);
+                ref byte dst = ref Unsafe.AddByteOffset(ref MemoryMarshal.GetArrayDataReference(destinationArray), (uint)destinationIndex * elementSize);
+
+                if (pMT->ContainsGCPointers)
+                    Buffer.BulkMoveWithWriteBarrier(ref dst, ref src, byteCount);
+                else
+                    SpanHelpers.Memmove(ref dst, ref src, byteCount);
+
+                // GC.KeepAlive(sourceArray) not required. pMT kept alive via sourceArray
+                return;
+            }
+
+            // If we were called from Array.ConstrainedCopy, ensure that the array copy
+            // is guaranteed to succeed.
+            if (reliable)
+                throw new ArrayTypeMismatchException(SR.ArrayTypeMismatch_ConstrainedCopy);
+
+            // Rare
+            CopySlow(sourceArray, sourceIndex, destinationArray, destinationIndex, length, assignType);
+        }
+
+        // Reliability-wise, this method will either possibly corrupt your
+        // instance & might fail when called from within a CER, or if the
```

Comment:
```suggestion
        // instance, or if the
```
We do not have CERs anymore